psm_utils/io/mzid.py

```diff
@@ -0,0 +1,248 @@
+            SpectrumIdentificationItem["PeptideEvidenceRef"]
+        )
+        # retention time is often missing from mzid
+        try:
+            rt = float(SpectrumIdentificationItem[self.searchengine_key_dict["rt_key"]])
+        except KeyError:
+            rt = float("nan")
+
+        psm = PeptideSpectrumMatch(
+            peptide=peptide,
+            spectrum_id=str(spectrum_title),
+            run=str(rawfile),
+            is_decoy=isdecoy,
+            score=float(
+                SpectrumIdentificationItem[self.searchengine_key_dict["score_key"]]
+            ),
+            precursor_charge=int(SpectrumIdentificationItem["chargeState"]),
+            # Or use calculatedMassToCharge?
+            precursor_mz=float(SpectrumIdentificationItem["experimentalMassToCharge"]),
+            # Retention time often not included in mzid output
+            retention_time=rt,
+            protein_list=protein_list,
+            source=self.source,
+            provenance_data=({f"{self.source}_filename": self.filename}),
+            metadata=self._get_searchengine_specific_metadata(
+                SpectrumIdentificationItem
+            ),
+        )
+        return psm
+
+    @staticmethod
+    def _parse_peptidoform(seq: str, modification_list: list[dict]):
+        """Return a peptido form"""
+        # TODO: not able to differentiate between c-terminal mod and mod on last aa
+
+        modification_lengths = 0
+        for mod in modification_list:
+            if mod["location"] == 0:
+                proforma_mod = f"[{mod['name']}]-"
+                seq = proforma_mod + seq
+
+            else:
+                proforma_mod = f"[{mod['name']}]"
+                seq = (
+                    seq[0 : mod["location"] + modification_lengths]
+                    + proforma_mod
+                    + seq[mod["location"] + modification_lengths :]
+                )
+
+            modification_lengths += len(proforma_mod)
+
+        return Peptidoform(seq)
+
+    def _infer_source(self):
+        """Get the source of the mzid file"""
+
+        mzid_xml = ET.parse(self.filename)
+        root = mzid_xml.getroot()
+        name_space = self._get_xml_namespace(root.tag)
+
+        return root.find(f".//{name_space}AnalysisSoftware").attrib["name"]
+
+    @staticmethod
+    def _get_xml_namespace(root_tag):
+        """Get the namespace of the xml root"""
+
+        m = re.match(r"\{.*\}", root_tag)
+        return m.group(0) if m else ""
+
+    @staticmethod
+    def _parse_PeptideEvidenceRef(PeptideEvidenceList: list[dict]):
+        """Parse peptide evidence References of PSM"""
+
+        isdecoy = PeptideEvidenceList[0]["isDecoy"]
+
+        protein_list = [
+            d["accession"] for d in PeptideEvidenceList if "accession" in d.keys()
+        ]
+
+        return isdecoy, protein_list
+
+    def _get_searchengine_specific_keys(self):
+        """Get searchengine specific keys"""
+        # TODO works for PEAKS pro?
```

Comment:
Still looking for an PEAKS pro file

Reply:
I think all PEAKS software will say "PEAKS Studio" as the test file now does. For now I put it up like this to just look for PEAKS in the now, if other names are used it will still take the right searchengine keys